rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,849 @@
+
+### Config Changes
+
+Encrypted session recording is a feature of the auth service and can be enabled
+using a new `encryption` section in the `session_recording_config` resource.
+
+```yaml
+# session_recording_config.yml
+kind: session_recording_config
+version: v2
+spec:
+  encryption:
+    # whether or not encryption should be enabled
+    enabled: true
+    # whether or not Teleport should manage the keys or simply consume them
+    manual_key_management:
+      enabled: true
+      # key labels Teleport will use to find active encryption keys when
+      # 'enabled' is true
+      active_keys:
+        - type: pkcs11
+          label: 'pkcs11-key-label'
+        - type: aws_kms
+          label: 'aws key ID'
+      # key labels Teleport will use to find rotated decryption keys when
+      # 'enabled' is true
+      rotated_keys: []
+```
+
+HSM integration is facilitated through the existing configuration
+options for setting up an HSM backed CA keystore through pkcs#11. Example
+configuration found [here](https://goteleport.com/docs/admin-guides/deploy-a-cluster/hsm/#step-25-configure-teleport).
+
+The `manual_key_management` configuration controls whether or not the wrapping
+keys used by `age` should be provisioned and managed by Teleport or by the
+configured keystore. When disabled, Teleport will provision, share, and rotate
+keys as necessary. When enabled, Teleport expects to have access to keys
+queryable using the key labels assigned to `active_keys` and `rotated_keys`.
+Any keys found will be used during encryption and decryption and Teleport will
+not attempt to provision new keys or facilitate rotation.
+
+These configuration options are also accessible in the `teleport.yml` file
+configuration under the `auth_service` section:
+
+```yaml
+# teleport.yml
+auth_service:
+  session_recording: node
+  session_recording_config:
+    encryption:
+      enabled: true
+      manual_key_management:
+        enabled: true
+        active_keys:
+          - type: aws_kms
+            label: 'aws key ID'
+        rotated_keys: []
+```
+
+### Protobuf Changes
+
+```proto
+// api/proto/teleport/legacy/types/types.proto
+
+// EncryptionKeyPair is a keypair used for encrypting and decrypting data.
+message EncryptionKeyPair {
+  // public_key is the public encryption key.
+  bytes public_key = 1 [(gogoproto.jsontag) = "public_key"];
+  // PrivateKey is the private decryption key.
+  bytes private_key = 2 [(gogoproto.jsontag) = "private_key"];
+  // PrivateKeyType is the type of the private_key.
+  PrivateKeyType private_key_type = 3 [(gogoproto.jsontag) = "private_key_type"];
+  // Hash is the hash function to use during encryption/decryption operations.
+  // It maps directly to the possible values of crypto.Hash in the go crypto
+  // package.
+  uint32 hash = 4 [(gogoproto.jsontag) = "hash"];
+}
+
+// AgeEncryptionKey is a PEM encoded RSA4096 public key used for encrypting
+// session recordings.
```

Comment:
Don't use PEM especially if it's stored in a `bytes` field, store the spki in DER form instead (which could likely not need anything more to specify that it's a RSA public key rather than something else).